README.md
# Timetables

![timetables](i/timetables.jpg)

### About

*A Perl script* to fetch and parse arbitrary HTML code
and to design and write custom timetable data in a proprietary file format loaded eventually in an indoor display.

### See [NOTES](NOTES.md).

(c) 2015 Pavel Jurca, University of Economics in Prague


changelog.txt
commit 8e6f19736d7add974d8492c6ec057c542973fbbd
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Wed Oct 19 12:37:34 2016 +0200

    diff between git version and live version on the system

commit 1cbb6df21f28500fb04759c509c752d5d7bb933e
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Wed Oct 19 12:22:29 2016 +0200

    added debug line for testing; calling 'open' needs to be commented out though

commit ee959d79583453d563f4b39b205df4e0b4eb0446
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Mon Oct 12 19:05:54 2015 +0200

    clarified

commit 080a08d7153b5b665d5257195e89d2e614ef92d3
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Mon Oct 12 15:34:27 2015 +0200

    pretty

commit af779dcbdcc56de350aec4c743d27f689208851f
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Tue Sep 29 08:53:36 2015 +0200

    's///' inside of 'map' destroying orig values

commit 770be09c4d568d6476ae1882d5441adf28843687
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Fri Sep 25 17:07:55 2015 +0200

    min 6 chars column width; fixed silly bug to get digits from string;

commit 385aae33c338c653467988329f4b9558b7f41640
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Tue Sep 22 10:37:53 2015 +0200

    increased panel delay to 51 sec

commit e862a8c9324b1c68f6e144a09780a678dc7c6115
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Tue Sep 22 10:36:10 2015 +0200

    changed '&&' for 'and' in flow control

commit b3c3051d148bcd4e12201b3f89738b32fc0f810c
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Thu Sep 17 14:02:20 2015 +0200

    Jižní Město

commit 0a3649f5672ea6b0cebd469fe94276ef92ce4d87
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Fri Sep 11 23:22:58 2015 +0200

    && over *and* in logical comparison

commit 2f001e65f46abfa4c97b1d16f6f83f9c4c658c0d
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Fri Sep 11 08:46:28 2015 +0200

    mea culpa: s/// in map changed obviously the original values too

commit 71264665861c49af52288d969ab8ca5774e8da96
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Thu Sep 10 18:29:17 2015 +0200

    fixed $prijm being undefined

commit 2725d1c226d18e617b349e717b5f55e248136f00
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Thu Sep 10 10:19:47 2015 +0200

    undefined regex match; the higher precedence || breaks a list-context

commit 5a141d566d5f7b243a31b6fa5ad42d803a345a6a
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Wed Sep 9 19:11:13 2015 +0200

    'Ž' to 'ž'

commit c0188ec3328e55368e4c55145feb2d2d0cae34f4
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Wed Sep 9 18:59:32 2015 +0200

    fixed map func param greediness

commit 93706c0572dc95ce02fddfa53e2349af809fb32f
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Wed Sep 9 17:55:55 2015 +0200

    changed 'z' to 'ž' in a charset table

commit 43a72a7e5ecbc4656054deeb7b69b88da98cb162
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Wed Sep 9 15:02:27 2015 +0200

    adjusted text output and delay; a better placeholder

commit 8e8300028e8d7554e3c5dc373b58b7187312a019
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Wed Sep 9 14:32:26 2015 +0200

    rozvrhy mimo pravidelny cas, a placeholder v pripade volneho dne

commit e6167c92ed8a91e6f2452dbb30c00bf8b58ed6ae
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Fri Sep 4 16:39:56 2015 +0200

    @

commit fbfdff4047773622af92a1fed56cdf7181fa66ab
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Fri Sep 4 14:27:27 2015 +0200

    @

commit f10e8f13b8e4ae1ad7725832fa47ff68c1029b86
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Fri Sep 4 10:55:00 2015 +0200

    formatted perldoc better

commit fe122503e7c19d7d6769a1db85ec7d04ba519d06
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Wed Sep 2 12:42:40 2015 +0200

    removed chomp call

commit 56dede62d97c263c712d67f5d38d332e41039477
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Wed Sep 2 08:36:21 2015 +0200

    prettified

commit 4976d9984c5d4b4ba61fdb937bb13090a2fb2685
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Tue Sep 1 16:58:39 2015 +0200

    I did it!

commit 67c58f6542555e2542eaadde7d7b200566e321b3
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Tue Sep 1 16:30:52 2015 +0200

    debug

commit b8aa85cb280ab950191dc4520331c1419dec47cd
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Tue Sep 1 16:15:38 2015 +0200

    added newline ahead of every $cas segment

commit 71606eeeb3cd6afda4f7ebd18cec12176c998ff6
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Tue Sep 1 15:55:22 2015 +0200

    done! v3

commit cc0a5c7464102e23f16557fbcb64d05692e17a1d
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Tue Sep 1 15:17:41 2015 +0200

    the @panel dereferenced right

commit 4033fd0abff90d9389e5eb35dcf8be92e514d331
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Tue Sep 1 14:05:04 2015 +0200

    done! v2

commit 53016351f64c10bbc69032ff2d26bd63d2ab86b7
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Tue Sep 1 11:15:28 2015 +0200

    znaky nad 127 ASCII nahrazeny '?'

commit 714adf66b74820e05c8d6faeb8dbdfdc13169df2
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Mon Aug 31 20:21:15 2015 +0200

    keeping it backward compatible ;)

commit 6c3d11ca1ebaeef49107fe624a522312929aace9
Author: Pavel Jurča <paveljurca@users.noreply.github.com>
Date:   Mon Aug 31 20:08:13 2015 +0200

    done!


timetables.pl
use strict;
use warnings;
use LWP::Simple;
use File::Spec;
#use Data::Dumper;
use utf8;

# CHANGELOG
# v1, 08/2015, Pavel Jurca
# v1.1, 09/2015, Pavel Jurca
#   — rezervace mimo pravidelny cas
#   — text 'REZERVACE' na miste IDENTu
#   — placeholder v pripade volneho dne

# TODO
# https://metacpan.org/pod/HTTP::Tiny#request 
# Accept-Language: cs


### ============== MAIN =============
use constant API
    # live
    => 'REDACTED'
    # debug
    # => 'http://localhost:8080/rozvrh.html?'
;
use constant ROZVRHY
    => qw( 7:30 9:15 11:00 12:45 14:30 16:15 18:00 )
;
my %ucebna = (
            # INSIS id
    'JM 357'  => 752,
    'JM 359'  => 746,
    'JM 360'  => 2281,
    'JM 361'  => 790,
    'JM 382'  => 700,
    'JM 352'  => 4870, # studovna
);

tisk_panely(rozvrhy(values %ucebna));
### ============ END ================



=head2 rozvrhy

stahne a zpracuje rozvrhy pro dnesni den

 Returns : % rozvrh_ref
 Args    : @ ucebny

=cut

sub rozvrhy {
    my @uceb_id = @_;

    my %rozvrh;
    for my $ucebna (@uceb_id) {
        # HTTP GET :iso-8859-2
        my $html = do {
            my $r;
            # 3 pokusy o spojeni
            for (1..3) {
                $r = get(API . $ucebna);
                last if $r;
            }

            $r;
        };
        die "(!) isis neodpovida\n" unless $html;

        for my $_rozvrh (_html($html)) {
            my @predmet = @{ $_rozvrh };

            my $od = shift @predmet;
            my $do = shift @predmet;
            # min. sirka sloupce
            my $predm = sprintf "%-6s " x @predmet, @predmet;

            # prazdne mezery pryc
            $predm =~ s/^\s+|\s+$//g;

            for my $cas (ROZVRHY) {
                my ($_od, $_do, $_cas) = map {
                    (my $num = $_) =~ s/\D//g;
                    $num;
                } ($od, $do, $cas);

                # ulozi i rezervace pres vice hodin
                $rozvrh{$cas}{$ucebna} = $predm
                    if ($_cas >= $_od and $_cas < $_do);
            }
        }
    }

    return \%rozvrh;
}

=head2 _html

vyjme udaje z HTML rozvrhu,
ocisti a filtruje konkretni sloupce

 Returns : @ predmety
 Args    : text 

=cut

sub _html {
    my $html = shift;

    # ucebna je cely den prazdna
    return if $html !~ /<thead>/;

    my (@predmety, @tr);
    for (split m|</td>|, $html) {
        # chomp;
        if (m|</tr| && @tr > 5) {

            # Vyucujici, jen prijmeni
            my ($prijm) = split(/,/, $tr[5], 2);
            ($prijm) = $prijm =~ /[ ]?(\S*)$/;

            # Kod
            $tr[3] = 'REZERVACE'
                if $tr[4] =~ s/(reservation|rezervace):&nbsp;//i;

            # TECHNICAL MAINTENANCE STAFF hack
            $tr[3] = "Technická údržba",
              $prijm = "" if $prijm =~ /REDACTED|REDACTED|REDACTED/;

            push @predmety, [
                              $tr[0], # Od
                              $tr[1], # Do
                              $tr[3], # Kod
                              $prijm, # Vyucujici
                          ];
            @tr = (); #flip-flop
        }

        my ($td) = m|>([^<]+)</small>$|;
        # no space, tab or newline
        $td =~ s/^\s+|\s+$//g if $td;

        push @tr, $td || ' ';
    }

    # rozvrh pro danou ucebnu
    return @predmety;
}

=head2 tisk_panely

zapise do souboru 0730.325, 0915.325,
1100.325 apod. obrazovky panelu

(!) cokoliv nad 127 ASCII je treba kodovat

 Args    : % rozvrh_ref

=cut

sub tisk_panely {
    my $rozvrh_ref = shift;

    my @panel;
    for (my $i = 0; $i < (ROZVRHY); $i++) {
        my $cas = (ROZVRHY)[$i];

        my @_panel;
        if (keys %{ $rozvrh_ref }) {

            @_panel = ' ' x 9 . "|od ${cas}h|";

            for my $uceb (reverse sort keys %ucebna) {
                my $predm = \$rozvrh_ref->{$cas}{ $ucebna{$uceb} };
                push @_panel, $uceb . '  ' . ( $$predm || 'n/a' );
            }
        }
        else {
            # CELY DEN VOLNO
            push @_panel,
                    '',
                    (map { ' ' x 6 . $_ } qw/Vysoká škola ekonomická/),
                    '',
                    ' ' x 6 . 'Jižní Město'
            ;
        }

        # panel je formou 2 subpanelu
        # po 5 udajich kazdy (viz pocet uceben)
        push @panel, [ @_panel ];

            # Probiha / Nasleduje
        if (@panel == 2) {
            SOUBOR:
            {
                (my $soubor = (ROZVRHY)[$i-$#panel]) =~ s/\D//g;
                $soubor = sprintf "%04s", $soubor;

                my $cesta = File::Spec->catpath(
                    'C:',
                    File::Spec->catdir('', 'PANEL', 'TEXTS'),
                    $soubor . '.325'
                );

                #========== zacatek SOUBORU ===========

                open(my $fh, '>', $cesta)
                    or die "(!) nelze zapsat do '$cesta'\n";

                # DEBUG
                # my $fh = *STDOUT;
                # LIVE
                binmode $fh, ':bytes';

                # kontrolni znaky ~ 1 radek
                print $fh
                    '33', # 3,3 mod animace
                    '55', # 55s prodleva
                    "\n"
                ;

                  #====== OBSAH panelu ======

                  # zahlavi ~ 1 radek
                  print $fh
                      # den   # mesic
                      "%d. \x{BC}\x{BD}.",
                      ' ' x 2,
                      zakoduj('VŠE v Praze'),
                      ' ' x 2,  # vteriny
                      "%H:%M:\x{B8}\x{B9}",
                      "\n"
                  ;

                  # telo ~ 14 radku
                  print $fh
                      zakoduj(
                          join "\n", map {
                              # $_ // ''
                              defined $_ ? $_ : ''
                          } (map @$_, @panel[0..$#panel])[0..13]
                      ),
                      "\n"
                  ;

                  #=========================

                # kontrolni znaky ~ 15 radku
                print $fh
                    "\x{0F}" x 30, "\n" for (1..15)
                ;

                close $fh;

                #=====================================

                # Nasleduje => Probiha
                shift @panel;

                # posledni rozvrh dne
                redo SOUBOR if $i + @panel == (ROZVRHY);
            }
        }
    }

    print "(i) zapsano\n";
}

=head2 zakoduj

zakoduje text dle sady
pro panel SPECTRUM

 Returns : text
 Args    : text

=cut

sub zakoduj {
    my $text = shift;

    # znakova sada SPECTRUM 
    my %sada = (
        'A'  => "\x{41}",
        'Á'  => "\x{80}",
        'Ä'  => "\x{8E}",
        'B'  => "\x{42}",
        'C'  => "\x{43}",
        'Č'  => "\x{AC}",
        'D'  => "\x{44}",
        'Ď'  => "\x{D2}",
        'E'  => "\x{45}",
        'É'  => "\x{90}",
        'Ě'  => "\x{B7}",
        'Ë'  => "\x{45}",
        'F'  => "\x{46}",
        'G'  => "\x{47}",
        'H'  => "\x{48}",
        'I'  => "\x{49}",
        'Í'  => "\x{D6}",
        'Ï'  => "\x{8B}",
        'J'  => "\x{4A}",
        'K'  => "\x{4B}",
        'L'  => "\x{4C}",
        'Ľ'  => "\x{6C}",
        'M'  => "\x{4D}",
        'N'  => "\x{4E}",
        'Ň'  => "\x{D5}",
        'O'  => "\x{4F}",
        'Ó'  => "\x{E0}",
        'Ö'  => "\x{99}",
        'P'  => "\x{50}",
        'Q'  => "\x{51}",
        'R'  => "\x{52}",
        'Ř'  => "\x{FC}",
        'S'  => "\x{53}",
        'Š'  => "\x{8D}",
        'T'  => "\x{54}",
        'Ť'  => "\x{9C}",
        'U'  => "\x{55}",
        'Ú'  => "\x{8F}",
        'Ů'  => "\x{8C}",
        'Ü'  => "\x{9A}",
        'V'  => "\x{56}",
        'W'  => "\x{57}",
        'X'  => "\x{58}",
        'Y'  => "\x{59}",
        'Ý'  => "\x{ED}",
        'Ž'  => "\x{A7}",
        'a'  => "\x{61}",
        'á'  => "\x{A0}",
        'ä'  => "\x{84}",
        'b'  => "\x{62}",
        'c'  => "\x{63}",
        'č'  => "\x{9F}",
        'd'  => "\x{64}",
        'ď'  => "\x{88}",
        'e'  => "\x{65}",
        'é'  => "\x{82}",
        'ě'  => "\x{D8}",
        'ë'  => "\x{89}",
        'f'  => "\x{66}",
        'g'  => "\x{67}",
        'h'  => "\x{68}",
        'i'  => "\x{69}",
        'í'  => "\x{A1}",
        'ï'  => "\x{8B}",
        'j'  => "\x{6A}",
        'k'  => "\x{6B}",
        'l'  => "\x{6C}",
        'ľ'  => "\x{4C}",
        'm'  => "\x{6D}",
        'n'  => "\x{6E}",
        'ň'  => "\x{E5}",
        'o'  => "\x{6F}",
        'ó'  => "\x{A2}",
        'ö'  => "\x{94}",
        'p'  => "\x{70}",
        'q'  => "\x{71}",
        'r'  => "\x{72}",
        'ř'  => "\x{FD}",
        's'  => "\x{73}",
        'š'  => "\x{E7}",
        't'  => "\x{74}",
        'ť'  => "\x{9B}",
        'u'  => "\x{75}",
        'ú'  => "\x{A3}",
        'ů'  => "\x{85}",
        'ü'  => "\x{81}",
        'v'  => "\x{76}",
        'w'  => "\x{77}",
        'x'  => "\x{78}",
        'y'  => "\x{79}",
        'ý'  => "\x{EC}",
        'ž'  => "\x{A7}",
    );

    return join '', map {
          $sada{$_} || (ord $_ < 128 ? $_ : '?')
      } split //, $text
    ;
}
